File: lua/high-scores/high-scores.lua
table.unpack = table.unpack or unpack;

local Self = {}

function Self.scores(self)
    return { table.unpack(self._scores) };
end

function Self.personal_best(self)
    return math.max(
        table.unpack(self._scores)
    )
end

function Self.personal_top_three(self)
    local best = self:scores()
    table.sort(best, function (a, b) return a > b end)
    return { best[1], best[2], best[3] }
end

function Self.latest(self)
    return self._scores[#self._scores]
end

local meta = {
    __index = Self,
    __newindex = function()
        error("Try to modify a immutable table.")
    end
}

return function(scores)
    assert(
        type(scores) == "table",
        "Invalid type for arguments scores, expects a table got" ..
        type(scores) .. "."
    )

    local self = {}

    self._scores = { table.unpack(scores) };
    setmetatable(self, meta)

    return self
end

File: java/high-scores/src/main/java/HighScores.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class HighScores {

    private final List<Integer> highScores;

    public HighScores(List<Integer> highScores) {
        this.highScores = new ArrayList<Integer>(highScores);
    }

    public List<Integer> scores() {
        return new ArrayList<Integer>(this.highScores);
    }

    public Integer latest() {
        return this.highScores
                .get(this.highScores.size() - 1);
    }

    public Integer personalBest() {
        return Collections
                .max(this.highScores);
    }

    public List<Integer> personalTopThree() {
        return this.highScores
                .stream()
                .sorted((a, b) -> b - a)
                .limit(3)
                .toList();
    }

}
